lib/std/crypto/salsa20.zig

```diff
@@ -0,0 +1,430 @@
+const std = @import("std");
+const crypto = std.crypto;
+const debug = std.debug;
+const math = std.math;
+const mem = std.mem;
+
+const Poly1305 = crypto.onetimeauth.Poly1305;
+const Blake2b = crypto.hash.blake2.Blake2b;
+const X25519 = crypto.dh.X25519;
+
+const Salsa20NonVecImpl = struct {
+    const BlockVec = [16]u32;
+
+    fn initContext(key: [8]u32, d: [4]u32) BlockVec {
+        const c = "expand 32-byte k";
+        const constant_le = comptime [4]u32{
+            mem.readIntLittle(u32, c[0..4]),
+            mem.readIntLittle(u32, c[4..8]),
+            mem.readIntLittle(u32, c[8..12]),
+            mem.readIntLittle(u32, c[12..16]),
+        };
+        return BlockVec{
+            constant_le[0], key[0],         key[1],         key[2],
+            key[3],         constant_le[1], d[0],           d[1],
+            d[2],           d[3],           constant_le[2], key[4],
+            key[5],         key[6],         key[7],         constant_le[3],
+        };
+    }
+
+    const QuarterRound = struct {
+        a: usize,
+        b: usize,
+        c: usize,
+        d: u6,
+    };
+
+    inline fn Rp(comptime a: usize, comptime b: usize, comptime c: usize, comptime d: u6) QuarterRound {
+        return QuarterRound{
+            .a = a,
+            .b = b,
+            .c = c,
+            .d = d,
+        };
+    }
+
+    inline fn salsa20Core(x: *BlockVec, input: BlockVec) void {
+        const arx_steps = comptime [_]QuarterRound{
+            Rp(4, 0, 12, 7),   Rp(8, 4, 0, 9),    Rp(12, 8, 4, 13),   Rp(0, 12, 8, 18),
+            Rp(9, 5, 1, 7),    Rp(13, 9, 5, 9),   Rp(1, 13, 9, 13),   Rp(5, 1, 13, 18),
+            Rp(14, 10, 6, 7),  Rp(2, 14, 10, 9),  Rp(6, 2, 14, 13),   Rp(10, 6, 2, 18),
+            Rp(3, 15, 11, 7),  Rp(7, 3, 15, 9),   Rp(11, 7, 3, 13),   Rp(15, 11, 7, 18),
+            Rp(1, 0, 3, 7),    Rp(2, 1, 0, 9),    Rp(3, 2, 1, 13),    Rp(0, 3, 2, 18),
+            Rp(6, 5, 4, 7),    Rp(7, 6, 5, 9),    Rp(4, 7, 6, 13),    Rp(5, 4, 7, 18),
+            Rp(11, 10, 9, 7),  Rp(8, 11, 10, 9),  Rp(9, 8, 11, 13),   Rp(10, 9, 8, 18),
+            Rp(12, 15, 14, 7), Rp(13, 12, 15, 9), Rp(14, 13, 12, 13), Rp(15, 14, 13, 18),
+        };
+        x.* = input;
```

Comment:
this suggests this function would be better returning a `BlockVec`

Reply:
The initial state and the first rounds should be overwritten as soon as possible. The intent of passing the state by address is to facilitate this.

Reply:
result location semantics should guarantee that to be the case.

Sorry that #2809 is incomplete :(